### PR TITLE
create an entity relationship diagrams of the schemas using erdantic

### PR DESCRIPTION
**Pull request by yves-renier**
Hello,

I wanted to have ERD diagram for my models so I implemented this.
I added an option model-erdantic-figure to switch it off (it is currently on by default, but feel free to change that).
I also added the documentation for the option.

### CODE REVIEW

**Comment by mansenfranzen** [comment]
Hi @yves-renier,

thanks for you PR! You did a great job diving deeply into `sphinx` and `autodoc_pydantic` internals ;-). 

I really like the idea of adding the possibility to show ERDs for pydantic models. I'm happy to support this PR.

So far, there remain some ToDos that we should address first:

- [ ] `autodoc_pydantic` will need to add `erdantic` as an dependency in its `pyproject.toml`. Otherwise, the docs build process runs into an import error.

- `erdantic` depends on `pygraphviz` and `graphviz` which will complicate the installation process of `autodoc_pydantic`.
  - `pygraphviz` does not offer wheels and [requires a C/C++ compiler](https://pygraphviz.github.io/). 
  - `graphviz` is not a python but external system dependency which can't be installed via pip.
  - Using `conda` here would simplify the process of setting up the doc build environment but I would rather prefer to keep it plain poetry/pip.
 - [ ] Hence, I suggest that `erdantic` should be an optional dependency. Accordingly, the ERD feature should be turned off by default.
 - [ ] If ERD is enabled, we should add a check to test against appropriate dependencies and provide a meaningful error message to inform users.
    - [ ] This will require a short documentation on how to properly set up `autodoc_pydantic` with `erdantic`, `pygraphviz` and `graphviz`.
- [ ] Since this is a great feature, I would very much like to see a dedicated example in the [example section](https://autodoc-pydantic.readthedocs.io/en/stable/users/examples.html).
- [ ] Currently, there are no tests for the new feature. This should be added. I can assist here since this is rather complicated due to the nature of `sphinx` tests. 

Please feel free to comment on my thoughts here. I'm happy to discuss them.

**Comment by yves-renier** [comment]
Hello, thanks for the support !
I just added `erdantic` as an optional dependency (`pip install autodoc_pydantic[erd]`) and turned off the ERD feature by default.
I will look to add an example, check and document the dependencies and try to setup the tests locally.

**Comment by mansenfranzen** [comment]
Thanks for continuing with the PR! 

If you get stuck or need any help, please don't hesitate to reach out.

**Comment by yves-renier** [comment]
About example, I already added one at `/docs/build/html/users/configuration.html#show-erdantic-figure`.
I will also enable the option for the [complete example](https://autodoc-pydantic.readthedocs.io/en/stable/users/examples.html#complete).

**Comment by yves-renier** [comment]
I just added a simple test.
@mansenfranzen, I believe I finished implementing your TODO.
I also modified the github actions to install erdantic and graphviz.

**Comment by mansenfranzen** [comment]
@yves-renier The PR looks great - well done!

I can confirm that tests pass and documentation is rendered properly. I have a few thoughts remaining that prevent me from merging, yet:

- I dislike `erdantic`'s watermark at the bottom of the diagram. It distracts. I would rather see it being removed by `erdantic` or by `autodoc_pydantic`. The latter option is of course less desirable. I opened an issue [here](https://github.com/drivendataorg/erdantic/issues/78).
- The resulting diagram can be improved if rendered as SVG instead of PNG. SVG is superior because it is a vector and additional information is shown upon hovering over attributes of pydantic models. `sphinx.ext.graphviz` already provides a configuration to switch to SVG, see [here](https://www.sphinx-doc.org/en/master/usage/extensions/graphviz.html#confval-graphviz_output_format). We could make this the default for `autodoc_pydantic`.
- While seeing the generated diagrams in action for the first time in the newly build docs, I think the diagrams can be visually very dominant. I think we can improve here to make them collapsable just like the [JSON schema](https://autodoc-pydantic.readthedocs.io/en/stable/users/configuration.html#show-schema-json).

I'm curious to get to know your thoughts on these points.

**Comment by yves-renier** [comment]
- The watermark made the test harder as `erdantic` version is mentioned in it which we don't know in advance. I ignored that part of the output in the tests I implemented. Having a customizable label would also make those kind of tests much cleaner. I totally agree with you that the best would be an option from `erdantic` as I would neither be comfortable with removing it from `autodoc_pydantic`.
- about SVG format for the diagram, I just tested and it seems to works. I modified docs/source/conf.py accordingly.
- about making the diagram collapsible, I have no problem about that, I will make it the default and add an other option `model-erdantic-figure-collapsed` to disable it.

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (mansenfranzen) [inline sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:360]: Sphinx' autodoc-directives already hold a reference for the object to be documented via `self.object`. Please use `self.object` and drop `_get_pydantic_class()` completely.

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (mansenfranzen) [inline sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:358]: Great usage of sphinx' logging and also adding a warning to the docs. However, I think it would be better to raise an exception here because if someone explicitly activates the ERD functionality, an error should be raised if it's not functioning. What do you think?

File: sphinxcontrib/autodoc_pydantic/directives/templates.py
Comment (mansenfranzen) [inline sphinxcontrib/autodoc_pydantic/directives/templates.py:11]: Thumbs up for reusing the template :-)

File: pyproject.toml
Comment (mansenfranzen) [inline pyproject.toml:1]: `erdantic= { version ="^0.5", optional = true }` needs to be added to `[tool.poetry.dependencies]`. Otherwise, the install with extras `erd` does not work (or at least did not work for me).

File: pyproject.toml
Comment (mansenfranzen) [inline pyproject.toml:53]: Let's add `erdantic` to the `dev`, `test` and `docs` extras because it will always be required from now. This will also simplify `tox.ini`.

File: tox.ini
Comment (mansenfranzen) [inline tox.ini:1]: `erd` extras can be removed if `erdantic` is added to all extras in `pyproject.toml`.

File: docs/source/users/configuration.rst
Comment (mansenfranzen) [inline docs/source/users/configuration.rst:217]: The reference is to the configuration section itself (collapsed). It should be `autodoc_pydantic_model_erdantic_figure` instead.

File: .github/workflows/tests.yml
Comment (mansenfranzen) [inline .github/workflows/tests.yml:1]: The tests are not running currently because the `runs-on` is outdated with `runs-on: ubuntu-18.04`. Let's change this to `ubuntu-22.04` and see if the runners pick up the tests again.

**Review by mansenfranzen — CHANGES_REQUESTED** [review]
As I said before, you've submitted great PR! I really appreciate the work you've put into the project.

I've added multiple small comments/recommendations. I hope they are self-explanatory. If not, please reach out.

I think we still need to include a short note/hint in the documentation about installing `erdantic` if the feature is enabled.

**Comment by mansenfranzen** [comment]
@yves-renier By the way, great work adding `model-erdantic-figure-collapsed` along with tests!

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (yves-renier) [reply sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py]: That's was my initial thought, but I was not sure you would be happy with exceptions being raised. As it does not seem to be a problem, I will do it that way instead.

**Comment by yves-renier** [comment]
all tests passed on my repo (see https://github.com/yves-renier/autodoc_pydantic/actions/runs/4978871857/jobs/8909737895 ) except for the uploading of the documentation as I did not configure the upload token.

**Comment by mansenfranzen** [comment]
@yves-renier Great work! I'm sorry for letting you wait so long. I was unfortunately busy otherwise.

Great job on adding the new sphinx versions to the test matrix while also fixing the corresponding sphinx compatibility issues. In fact, your PR is already worth multiple PRs because you've also fixed and updated the CI pipeline, too. 

There only minor points that I would like to comment on. I can happily finish it from here if you would allow me to commit on your fork ([see here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)). Otherwise, feel free to finish it yourself ;-).

After this PR is merged, I would be very curious to understand how you approached this PR. More specifically, did you use the developer's documentation? If so, was it helpful or did you miss something? In addition, you are by far the most involved committer towards the code base. Your feedback regarding design, structure and complexity would be highly appreciated. For example, what parts of the code base where not intuitive? What make you frown?

File: docs/source/users/configuration.rst
Comment (mansenfranzen) [inline docs/source/users/configuration.rst:218]: I feel we should also include a short warning here about the `graphviz` dependency. You've added one in the user's installation section, but most will probably won't read it anyway and rather come here to understand how to enable the feature.

File: pyproject.toml
Comment (mansenfranzen) [inline pyproject.toml:1]: Let's drop support for python 3.6 altogether. I think we can safely use 

```
[tool.poetry.dependencies]
python = ">=3.7.1,<4.0.0"
```

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (mansenfranzen) [inline sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:290]: In my opinion, we can move `self.pydantic.options.is_true("erdantic-figure-collapsed", True)` into the `self.add_erdantic_figure`. This would remove the `if` condition here and also would make the `collapsed` argument of `self.add_erdantic_figure` obsolete. Overall, that would be a minor reduction of complexity.

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (mansenfranzen) [inline sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:370]: This block appears twice since we have the second feature that uses collapsible blocks (`autodoc_pydantic_collapsable_json`). To prevent code duplication, I suggest we use a function like this located in `sphinxcontrib/autodoc_pydantic/directives/templates.py`:
```python
def to_collapsable(lines: list[str], title, css_class) -> list[str]:
    """Place given lines into a collapsable HTML block.
    
    Parameters
    ----------
    lines: list
        The actual content, that should be placed into a collapsable block.
    title: str
        The name of the collapsable block title.
    css_class: str
        Name of the css class for the collapsable block.
        
    """
    
    lines = "\n".join(lines)
    lines = TPL_COLLAPSE.format(
        lines=lines,
        summary=title,
        details_class=css_class
    )
    return lines.split("\n")
```

Then, we could change the marked code to:

```python
if collapsed:
  lines = to_collapsable(
      lines=lines,
      title="Show Entity Relationship Diagram",
      css_class="autodoc_pydantic_collapsable_erd"
  )
```

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (mansenfranzen) [inline sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py:362]: For my taste, I would like to make this part a bit more concise:

```python
figure_dot = erd.to_dot(self.object).replace('\t', '   ').split('\n')
lines_dot = ['   ' + line for line in figure_dot]
lines = [".. graphviz::", ""] + lines_dot + [""]
```

File: tests/compatibility.py
Comment (mansenfranzen) [inline tests/compatibility.py:146]: Could you please provide a simple doc string with an example of input -> output? This would greatly help to understand what this function does. Moreover, the function name itself is rather generic and requires some guessing.

**Review by mansenfranzen — COMMENTED** [review]
As I said before, excellent work! I've added some more comments but this PR would already qualify for merging.

In the end, we should not forget about the changelog. Surely I can take care of this if you want.

**Comment by yves-renier** [comment]
Sorry, I totally missed the GH messages. Now I have seen it I will put back some effort into it.

>  if you would allow me to commit on your fork

I have done that during the PR creation, let me know if it does not work.

I will now go through your comments, then I will give you my feedback on the project.

File: docs/source/users/configuration.rst
Comment (yves-renier) [reply docs/source/users/configuration.rst]: implemented in 803e7c3

File: pyproject.toml
Comment (yves-renier) [reply pyproject.toml:1]: fixed by 3e70a68

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (yves-renier) [reply sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py]: fixed by ff24325

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (yves-renier) [reply sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py]: fixed in 67475f0

File: sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py
Comment (yves-renier) [reply sphinxcontrib/autodoc_pydantic/directives/autodocumenters.py]: done in 25af89a

File: tests/compatibility.py
Comment (yves-renier) [reply tests/compatibility.py]: done in 1b7df9b

**Comment by yves-renier** [comment]
> In the end, we should not forget about the changelog. Surely I can take care of this if you want.

My last commit tried to address that. I tried to follow the format of the previous entries, but fell free to modify anything as you wish.

**Comment by mansenfranzen** [comment]
> adding a link in the Documentation section of the Readme.md would not hurt.

Agreed - I will incorporate that in a separate release.

> One last thing, I did not manage to run sphinx in debug mode in my IDE (pycharm) which made the debugging of issues quite painful. If you know a way to achieve that, it could be a nice tip to add in the developer's documentation.

What was the error you were running into? Without pycharm's debugger I would be lost.

**Comment by mansenfranzen** [comment]
I made several small modifications:

- Added a separate `extras` in pyproject.toml for `erdantic` as you've initially implemented it. It makes the user install and the CI setup with erdantic as a dependency both more explicit.
- Added a separate test running the test suite without installing erdantic since there were no tests without erdantic as a dependency anymore. Since we using tox, the package is build from scratch and installed in clean environment which helps to ensure that build/install process works fine.
- Updates docs/readme with versions/badges.

**Comment by mansenfranzen** [comment]
Once again, thanks for all the great work you've put into this PR! It was a pleasure working with you :-).

**Pull request merged**
